File: app/src/main/java/netpro/keyTransmitter/AddKeyDIalogFragment.java
package netpro.keyTransmitter;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;

import java.util.ArrayList;
import java.util.List;

import static android.view.View.GONE;

public class AddKeyDIalogFragment extends android.support.v4.app.DialogFragment {

    private OnKeyGeneratedListener listener;
    private List<View> keyCodeViewList = new ArrayList<>();

    public static AddKeyDIalogFragment newInstance(int emptySpace) {
        AddKeyDIalogFragment fragment = new AddKeyDIalogFragment();
        Bundle args = new Bundle();
        args.putInt("emptySpace", emptySpace);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public Dialog onCreateDialog(final Bundle savedInstanceState) {
        final View view = View.inflate(getActivity(), R.layout.dialog_fragment_add_key, null);

        final TextInputLayout intervalLayout = (TextInputLayout) view.findViewById(R.id.intervalLayout);
        intervalLayout.setVisibility(GONE);
        intervalLayout.setErrorEnabled(true);
        intervalLayout.setError("必須");
        intervalLayout.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                try {
                    Integer.parseInt(editable.toString());
                } catch (NumberFormatException e) {
                    intervalLayout.setErrorEnabled(true);
                    intervalLayout.setError("必須");
                    return;
                }
                intervalLayout.setErrorEnabled(false);
            }
        });

        final TextInputLayout nameTil = (TextInputLayout) view.findViewById(R.id.nameTextInputLayout);
        nameTil.setErrorEnabled(true);
        nameTil.setError("必須");
        nameTil.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                Spinner keyTypeSpinner = (Spinner) view.findViewById(R.id.keyTypeSpinner);
                String str = (String) keyTypeSpinner.getSelectedItem();
                Key.Type type = Key.Type.toType(str);
                if (type != Key.Type.EMPTY && editable.length() == 0) {
                    nameTil.setErrorEnabled(true);
                    nameTil.setError("必須");
                } else {
                    nameTil.setErrorEnabled(false);
                }
            }
        });

        final TextInputLayout descriptionTil = (TextInputLayout) view.findViewById(R.id.descriptionTextInputLayout);
        descriptionTil.setErrorEnabled(true);
        descriptionTil.setError("必須");
        descriptionTil.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                Spinner keyTypeSpinner = (Spinner) view.findViewById(R.id.keyTypeSpinner);
                String str = (String) keyTypeSpinner.getSelectedItem();
                Key.Type type = Key.Type.toType(str);
                if (type != Key.Type.EMPTY && editable.length() == 0) {
                    descriptionTil.setErrorEnabled(true);
                    descriptionTil.setError("必須");
                } else {
                    descriptionTil.setErrorEnabled(false);
                }
            }
        });

        final LinearLayout addKeyLayout = (LinearLayout) view.findViewById(R.id.addKeyLayout);

        final Spinner columnCountSpinner = (Spinner) view.findViewById(R.id.columnCountSpinner);
        Integer[] columnCounts = {1, 2, 3, 4, 5};
        ArrayAdapter<Integer> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, columnCounts);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        columnCountSpinner.setAdapter(adapter);
        columnCountSpinner.setSelection(0);

        final Spinner rowCountSpinner = (Spinner) view.findViewById(R.id.rowCountSpinner);
        Integer[] rowCounts = {1, 2, 3, 4};
        adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, rowCounts);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        rowCountSpinner.setAdapter(adapter);
        rowCountSpinner.setSelection(0);

        final Spinner keyTypeSpinner = (Spinner) view.findViewById(R.id.keyTypeSpinner);
        Key.Type[] keyTypes = Key.Type.values();
        String[] keyDescriptions = new String[keyTypes.length];
        for (int i = 0; i < keyDescriptions.length; i++) {
            keyDescriptions[i] = keyTypes[i].getDescription();
        }

        ArrayAdapter<String> keyAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, keyDescriptions);
        keyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        keyTypeSpinner.setAdapter(keyAdapter);
        keyTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                Spinner spinner = (Spinner) adapterView;
                String description = (String) spinner.getSelectedItem();
                Key.Type type = Key.Type.toType(description);

                if (type == Key.Type.EMPTY) {
                    //名前と説明を必須にしない
                    nameTil.setErrorEnabled(false);
                    descriptionTil.setErrorEnabled(false);
                    //キーを追加できなくする
                    addKeyLayout.setVisibility(GONE);
                } else {
                    if (nameTil.getEditText().getText().length() == 0) {
                        nameTil.setErrorEnabled(true);
                        nameTil.setError("必須");
                    }
                    if (descriptionTil.getEditText().getText().length() == 0) {
                        descriptionTil.setErrorEnabled(true);
                        descriptionTil.setError("必須");
                    }
                    addKeyLayout.setVisibility(View.VISIBLE);
                }

                if (type == Key.Type.PRESSING) {
                    //キー送信間隔入力ボックスを表示
                    intervalLayout.setVisibility(View.VISIBLE);
                } else {
                    intervalLayout.setVisibility(GONE);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
        keyTypeSpinner.setSelection(0);

        final LinearLayout linearLayout = (LinearLayout) view.findViewById(R.id.keyCodesLayout);

        final Button addButton = (Button) view.findViewById(R.id.addButton);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                //Spinnerに入れるキーコードの種類を選択させるダイアログを生成
                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                final String[] items = {"アルファベット", "数字", "制御キー", "ファンクションキー", "記号"};
                builder.setItems(items, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        View addKeyCodeView = View.inflate(getActivity(), R.layout.layout_add_key_code, null);

                        Spinner spinner = (Spinner) addKeyCodeView.findViewById(R.id.keyCodeSpinner);
                        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item);
                        //選択結果に応じてSpinnerの選択肢を変える
                        switch (which) {
                            //アルファベット
                            case 0:
                                final int ALPHABET_SIZE = 'Z' - 'A';
                                char alphabet = 'A';
                                for (int i = 0; i <= ALPHABET_SIZE; i++) {
                                    adapter.add(String.valueOf(alphabet++));
                                }
                                break;
                            //数字
                            case 1:
                                for (int i = 0; i < 10; i++) {
                                    adapter.add(String.valueOf(i));
                                }
                                break;
                            //制御キー
                            case 2:
                                adapter.addAll("BackSpace", "Enter", "Shift", "Ctrl", "Alt", "Pause", "Space", "PageUp" + "PageDown", "End", "Home", "←", "↑", "→", "↓", "PrintScreen", "Insert", "Delete", "Win", "NumLock", "ScrollLock", "Esc", "Tab");
                                break;
                            //ファンクションキー
                            case 3:
                                for (int i = 1; i <= 12; i++) {
                                    adapter.add("F" + String.valueOf(i));
                                }
                                break;
                            //記号
                            case 4:
                                adapter.addAll(":*", ";+", ",<", "-=", ".>", "/?", "@`", "[{", "\\|", "]}", "^~", "\\_");
                                break;
                        }
                        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                        spinner.setAdapter(adapter);
                        spinner.setSelection(0);

                        Button removeButton = (Button) addKeyCodeView.findViewById(R.id.removeButton);
                        removeButton.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                ViewGroup parent = (ViewGroup) view.getParent();
                                keyCodeViewList.remove(parent);
                                parent.removeAllViews();
                            }
                        });
                        linearLayout.addView(addKeyCodeView);
                        keyCodeViewList.add(addKeyCodeView);
                    }
                });
                builder.show();
            }
        });

        Button createButton = (Button) view.findViewById(R.id.createButton);
        createButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view2) {
                EditText editName = (EditText) view.findViewById(R.id.name);
                String name = editName.getText().toString();

                EditText editDescription = (EditText) view.findViewById(R.id.description);
                String description = editDescription.getText().toString();

                int columnCount = (int) columnCountSpinner.getSelectedItem();
                int rowCount = (int) rowCountSpinner.getSelectedItem();

                Key.Type type = Key.Type.toType((String) keyTypeSpinner.getSelectedItem());
                Key key = new EmptyKey();
                switch (type) {
                    case RELEASED:
                        key = new NormalKey(columnCount, rowCount, name, description, type);
                        break;
                    case LONGPRESS:
                        key = new LongPressKey(columnCount, rowCount, name, description, type);
                        break;
                    case PRESSING:
                        EditText editInterval = (EditText) view.findViewById(R.id.interval);
                        int interval;
                        try {
                            interval = Integer.parseInt(editInterval.getText().toString());
                        } catch (NumberFormatException e) {
                            showErrorDialog("入力が不完全です");
                            return;
                        }
                        key = new PressingKey(columnCount, rowCount, name, description, type, interval);
                        break;
                    case EMPTY:
                        key = new EmptyKey(columnCount, rowCount, name, description, type);
                        break;
                }

                if (!(key instanceof EmptyKey)) {
                    if (name.length() == 0 || description.length() == 0) {
                        showErrorDialog("入力が不完全です");
                        return;
                    }

                    if (keyCodeViewList.size() <= 0) {
                        showErrorDialog("少なくとも1つの入力キーが必要です");
                        return;
                    }

                    for (View v : keyCodeViewList) {
                        Spinner spinner = (Spinner) v.findViewById(R.id.keyCodeSpinner);
                        key.addKeyCode((String) spinner.getSelectedItem());
                    }
                }

                if (columnCount * rowCount > getArguments().getInt("emptySpace")) {
                    int shortage = columnCount * rowCount - getArguments().getInt("emptySpace");
                    showErrorDialog("スペースが " + shortage + "足りません");
                    return;
                }

                listener.onKeyGenerated(key);
                dismiss();
            }
        });

        Button cacelButton = (Button) view.findViewById(R.id.cancelButton);
        cacelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
            }
        });

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("キー生成").setView(view);
        return builder.create();
    }

    private void showErrorDialog(String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle("エラー").setMessage(message).setPositiveButton("確認", null).show();
    }

    @Override
    public void onResume() {
        super.onResume();

        Dialog dialog = getDialog();
        WindowManager.LayoutParams layoutParams = dialog.getWindow().getAttributes();

        //display metricsでdpのもと(?)を作る
        DisplayMetrics metrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);

        //LayoutParamsにdpを計算して適用(今回は横幅300dp)(※metrics.scaledDensityの返り値はfloat)
        float dialogWidth = 300 * metrics.scaledDensity;
        layoutParams.width = (int) dialogWidth;

        //LayoutParamsをセットする
        dialog.getWindow().setAttributes(layoutParams);
    }


    @Override
    public void onPause() {
        super.onPause();
        dismiss();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnKeyGeneratedListener) {
            listener = (OnKeyGeneratedListener) context;
        } else {
            throw new RuntimeException(context.toString() + " must implement OnKeyGeneratedListener");
        }
    }

    public interface OnKeyGeneratedListener {
         void onKeyGenerated(Key key);
    }

}

File: app/src/main/java/netpro/keytransmitter/AddKeyDialogFragment.kt
package netpro.keytransmitter

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.support.v7.app.AlertDialog
import android.text.Editable
import android.text.TextWatcher
import android.util.DisplayMetrics
import android.view.View
import android.view.View.GONE
import android.view.ViewGroup
import android.widget.*
import java.util.*

class AddKeyDialogFragment : android.support.v4.app.DialogFragment() {

    private var listener: OnKeyGeneratedListener? = null
    private val keyCodeViewList = ArrayList<View>()
    private val rightKeyCodeViewList = ArrayList<View>()
    private val leftKeyCodeViewList = ArrayList<View>()
    private val flickUpKeyCodeViewList = ArrayList<View>()
    private val flickDownKeyCodeViewList = ArrayList<View>()
    private val flickRightKeyCodeViewList = ArrayList<View>()
    private val flickLeftKeyCodeViewList = ArrayList<View>()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = View.inflate(activity, R.layout.dialog_fragment_add_key, null)

        val intervalLayout = view.findViewById(R.id.intervalLayout) as TextInputLayout
        intervalLayout.visibility = GONE
        intervalLayout.isErrorEnabled = true
        intervalLayout.error = "必須"
        intervalLayout.editText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun afterTextChanged(editable: Editable) {
                try {
                    Integer.parseInt(editable.toString())
                } catch (e: NumberFormatException) {
                    intervalLayout.isErrorEnabled = true
                    intervalLayout.error = "必須"
                    return
                }

                intervalLayout.isErrorEnabled = false
            }
        })

        val adjustTextInputLayout = view.findViewById(R.id.text_input_layout_adjust) as TextInputLayout
        adjustTextInputLayout.visibility = GONE
        adjustTextInputLayout.isErrorEnabled = true
        adjustTextInputLayout.error = "必須"
        adjustTextInputLayout.editText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun afterTextChanged(editable: Editable) {
                try {
                    Integer.parseInt(editable.toString())
                } catch (e: NumberFormatException) {
                    adjustTextInputLayout.isErrorEnabled = true
                    adjustTextInputLayout.error = "必須"
                    return
                }

                adjustTextInputLayout.isErrorEnabled = false
            }
        })

        val descriptionTil = view.findViewById(R.id.descriptionTextInputLayout) as TextInputLayout
        descriptionTil.isErrorEnabled = true
        descriptionTil.error = "必須"
        descriptionTil.editText!!.addTextChangedListener(object:TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun afterTextChanged(editable: Editable) {
                val keyTypeSpinner = view.findViewById(R.id.keyTypeSpinner) as Spinner
                val str = keyTypeSpinner.selectedItem as String
                val type = Key.Type.toType(str)
                if (type != Key.Type.EMPTY && editable.length == 0) {
                    descriptionTil.isErrorEnabled = true
                    descriptionTil.error = "必須"
                } else {
                    descriptionTil.isErrorEnabled = false
                }
            }
        })

        val addKeyLayout = view.findViewById(R.id.addKeyLayout) as LinearLayout
        val addRightKeyLayout = view.findViewById(R.id.addRightKeyLayout) as LinearLayout
        val addLeftKeyLayout = view.findViewById(R.id.addLeftKeyLayout) as LinearLayout
        val addFlickUpKeyLayout = view.findViewById(R.id.addFlickUpKeyLayout) as LinearLayout
        val addFlickDownKeyLayout = view.findViewById(R.id.addFlickDownKeyLayout) as LinearLayout
        val addFlickRightKeyLayout = view.findViewById(R.id.addFlickRightKeyLayout) as LinearLayout
        val addFlickLeftKeyLayout = view.findViewById(R.id.addFlickLeftKeyLayout) as LinearLayout

        val columnCountSpinner = view.findViewById(R.id.columnCountSpinner) as Spinner
        val columnCounts = arrayOf(1, 2, 3, 4, 5)
        var adapter = ArrayAdapter(context, android.R.layout.simple_spinner_item, columnCounts)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        columnCountSpinner.adapter = adapter
        columnCountSpinner.setSelection(0)

        val rowCountSpinner = view.findViewById(R.id.rowCountSpinner) as Spinner
        val rowCounts = arrayOf(1, 2, 3, 4)
        adapter = ArrayAdapter(context, android.R.layout.simple_spinner_item, rowCounts)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        rowCountSpinner.adapter = adapter
        rowCountSpinner.setSelection(0)

        val keyTypeSpinner = view.findViewById(R.id.keyTypeSpinner) as Spinner
        val keyTypes = Key.Type.values()
        val keyDescriptions = arrayOfNulls<String>(keyTypes.size)
        for (i in keyDescriptions.indices) {
            keyDescriptions[i] = keyTypes[i].description
        }

        val keyAdapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item, keyDescriptions)
        keyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        keyTypeSpinner.adapter = keyAdapter
        keyTypeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                val spinner = adapterView as Spinner
                val description = spinner.selectedItem as String
                val type = Key.Type.toType(description)

                if (type == Key.Type.EMPTY) {
                    //名前と説明を必須にしない
                    descriptionTil.isErrorEnabled = false;
                    //キーを追加できなくする
                    addKeyLayout.visibility = GONE
                } else {
                    if (descriptionTil.editText!!.text.length == 0) {
                        descriptionTil.isErrorEnabled = true;
                        descriptionTil.error = "必須";
                    }

                    addKeyLayout.visibility = View.VISIBLE
                }

                if (type == Key.Type.PRESSING) {
                    //キー送信間隔入力ボックスを表示
                    intervalLayout.visibility = View.VISIBLE
                } else {
                    intervalLayout.visibility = GONE
                }

                if (type == Key.Type.KNOB) {
                    addKeyLayout.visibility = GONE
                    addRightKeyLayout.visibility = View.VISIBLE
                    addLeftKeyLayout.visibility = View.VISIBLE
                } else {
                    addRightKeyLayout.visibility = View.GONE
                    addLeftKeyLayout.visibility = View.GONE
                }

                if (type == Key.Type.FLICK) {
                    addKeyLayout.visibility = GONE
                    addFlickUpKeyLayout.visibility = View.VISIBLE
                    addFlickDownKeyLayout.visibility = View.VISIBLE
                    addFlickRightKeyLayout.visibility = View.VISIBLE
                    addFlickLeftKeyLayout.visibility = View.VISIBLE
                    adjustTextInputLayout.visibility = View.VISIBLE
                } else {
                    addFlickUpKeyLayout.visibility = View.GONE
                    addFlickDownKeyLayout.visibility = View.GONE
                    addFlickRightKeyLayout.visibility = View.GONE
                    addFlickLeftKeyLayout.visibility = View.GONE
                    adjustTextInputLayout.visibility = GONE
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {

            }
        }
        keyTypeSpinner.setSelection(0)

        val linearLayout = view.findViewById(R.id.keyCodesLayout) as LinearLayout
        val addButton = view.findViewById(R.id.addButton) as Button
        initKeyCodesLayout(linearLayout, addButton, keyCodeViewList)

        val rightKeyCodesLayout = view.findViewById(R.id.rotate_right_keyCodesLayout) as LinearLayout
        val addRightButton = view.findViewById(R.id.button_add_rotate_right_key) as Button
        initKeyCodesLayout(rightKeyCodesLayout, addRightButton, rightKeyCodeViewList)

        val leftKeyCodesLayout = view.findViewById(R.id.rotate_left_keyCodesLayout) as LinearLayout
        val addLeftButton = view.findViewById(R.id.button_add_rotate_left_key) as Button
        initKeyCodesLayout(leftKeyCodesLayout, addLeftButton, leftKeyCodeViewList)

        val flickUpKeyCodesLayout = view.findViewById(R.id.flick_up_keyCodesLayout) as LinearLayout
        val addFlickUpButton = view.findViewById(R.id.button_add_flick_up_key) as Button
        initKeyCodesLayout(flickUpKeyCodesLayout, addFlickUpButton, flickUpKeyCodeViewList)

        val flickDownKeyCodesLayout = view.findViewById(R.id.flick_down_keyCodesLayout) as LinearLayout
        val addFlickDownButton = view.findViewById(R.id.button_add_flick_down_key) as Button
        initKeyCodesLayout(flickDownKeyCodesLayout, addFlickDownButton, flickDownKeyCodeViewList)

        val flickRightKeyCodesLayout = view.findViewById(R.id.flick_right_keyCodesLayout) as LinearLayout
        val addFlickRightButton = view.findViewById(R.id.button_add_flick_raight_key) as Button
        initKeyCodesLayout(flickRightKeyCodesLayout, addFlickRightButton, flickRightKeyCodeViewList)

        val flickLeftKeyCodesLayout = view.findViewById(R.id.flick_left_keyCodesLayout) as LinearLayout
        val addFlickLeftButton = view.findViewById(R.id.button_add_flick_left_key) as Button
        initKeyCodesLayout(flickLeftKeyCodesLayout, addFlickLeftButton, flickLeftKeyCodeViewList)

        val createButton = view.findViewById(R.id.createButton) as Button
        createButton.setOnClickListener(View.OnClickListener {
            val editDescription = view.findViewById(R.id.description) as EditText
            val description = editDescription.text.toString()

            val columnCount = columnCountSpinner.selectedItem as Int
            val rowCount = rowCountSpinner.selectedItem as Int

            val type = Key.Type.toType(keyTypeSpinner.selectedItem as String)
            var key: Key = EmptyKey()
            when (type) {
                Key.Type.RELEASED -> key = NormalKey(columnCount, rowCount, description, type)
                Key.Type.LONGPRESS -> key = LongPressKey(columnCount, rowCount, description, type)
                Key.Type.PRESSING -> {
                    val editInterval = view.findViewById(R.id.interval) as EditText
                    val interval: Int
                    try {
                        interval = Integer.parseInt(editInterval.text.toString())
                    } catch (e: NumberFormatException) {
                        showErrorDialog("キー送信間隔を入力してください")
                        return@OnClickListener
                    }

                    key = PressingKey(columnCount, rowCount, description, type, interval.toLong())
                }
                Key.Type.KNOB -> key = ControlKnob(columnCount, rowCount, description, type)
                Key.Type.FLICK -> {
                    val adjustEditText = view.findViewById(R.id.edit_text_adjust) as EditText
                    val adjust: Int
                    try {
                        adjust = Integer.parseInt(adjustEditText.text.toString())
                    } catch (e: NumberFormatException) {
                        showErrorDialog("フリック距離を入力してください")
                        return@OnClickListener
                    }

                    key = FlickKey(columnCount, rowCount, description, type, adjust)
                }
                Key.Type.EMPTY -> key = EmptyKey(columnCount, rowCount, description, type)
            }

            if (key !is EmptyKey) {
                if (description.length == 0) {
                    showErrorDialog("入力が不完全です");
                    return@OnClickListener
                }

                if (key is ControlKnob) {
                    val rightKeyStrList = ArrayList<String>()
                    for (v in rightKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        rightKeyStrList.add(spinner.selectedItem as String)
                    }
                    val leftKeyStrList = ArrayList<String>()
                    for (v in leftKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        leftKeyStrList.add(spinner.selectedItem as String)
                    }
                    /*
                        if (rightKeyStrList.isEmpty() || leftKeyStrList.isEmpty()) {
                            showErrorDialog("少なくとも1つの入力キーが必要です");
                            return;
                        }
                        */
                    key.rotateRightKeyCodeList = rightKeyStrList
                    key.rotateRightKeyCodeList = leftKeyStrList
                } else if (key is FlickKey) {
                    val flickUpKeyStrList = ArrayList<String>()
                    for (v in flickUpKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        flickUpKeyStrList.add(spinner.selectedItem as String)
                    }
                    val flickDownKeyStrList = ArrayList<String>()
                    for (v in flickDownKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        flickDownKeyStrList.add(spinner.selectedItem as String)
                    }
                    val flickRightKeyStrList = ArrayList<String>()
                    for (v in flickRightKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        flickRightKeyStrList.add(spinner.selectedItem as String)
                    }
                    val flickLeftKeyStrList = ArrayList<String>()
                    for (v in flickLeftKeyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        flickLeftKeyStrList.add(spinner.selectedItem as String)
                    }
                    /*
                        if (rightKeyStrList.isEmpty() || leftKeyStrList.isEmpty()) {
                            showErrorDialog("少なくとも1つの入力キーが必要です");
                            return;
                        }
                        */
                    key.flickUpKeyStrList = flickUpKeyStrList
                    key.flickDownKeyStrList = flickDownKeyStrList
                    key.flickRightKeyStrList = flickRightKeyStrList
                    key.flickLeftKeyStrList = flickLeftKeyStrList
                } else {

                    /*
                        if (keyCodeViewList.size() <= 0) {
                            showErrorDialog("少なくとも1つの入力キーが必要です");
                            return;
                        }
                        */

                    for (v in keyCodeViewList) {
                        val spinner = v.findViewById(R.id.keyCodeSpinner) as Spinner
                        key.addKeyCode(spinner.selectedItem as String)
                    }
                }
            }

            if (columnCount * rowCount > arguments.getInt("emptySpace")) {
                val shortage = columnCount * rowCount - arguments.getInt("emptySpace")
                showErrorDialog("スペースが " + shortage + "足りません")
                return@OnClickListener
            }

            listener!!.onKeyGenerated(key)
            dismiss()
        })

        val cacelButton = view.findViewById(R.id.cancelButton) as Button
        cacelButton.setOnClickListener { dismiss() }

        val builder = AlertDialog.Builder(activity)
        builder.setTitle("キー生成").setView(view)
        return builder.create()
    }

    private fun showErrorDialog(message: String) {
        val builder = AlertDialog.Builder(context)
        builder.setTitle("エラー").setMessage(message).setPositiveButton("確認", null).show()
    }

    override fun onResume() {
        super.onResume()

        val dialog = dialog
        val layoutParams = dialog.window!!.attributes

        //display metricsでdpのもと(?)を作る
        val metrics = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(metrics)

        //LayoutParamsにdpを計算して適用(今回は横幅300dp)(※metrics.scaledDensityの返り値はfloat)
        val dialogWidth = 300 * metrics.scaledDensity
        layoutParams.width = dialogWidth.toInt()

        //LayoutParamsをセットする
        dialog.window!!.attributes = layoutParams
    }

    override fun onPause() {
        super.onPause()
        dismiss()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnKeyGeneratedListener) {
            listener = context as OnKeyGeneratedListener?
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnKeyGeneratedListener")
        }
    }

    interface OnKeyGeneratedListener {
        fun onKeyGenerated(key: Key)
    }

    private fun initKeyCodesLayout(keyCodesLayout: LinearLayout, addKeyCodeButton: Button, keyCodeViewList: MutableList<View>) {
        addKeyCodeButton.setOnClickListener {
            //Spinnerに入れるキーコードの種類を選択させるダイアログを生成
            val builder = AlertDialog.Builder(context)
            val items = arrayOf("アルファベット", "数字", "制御キー", "ファンクションキー", "特殊")
            builder.setItems(items) { _, which ->
                val addKeyCodeView = View.inflate(activity, R.layout.layout_add_key_code, null)

                val spinner = addKeyCodeView.findViewById(R.id.keyCodeSpinner) as Spinner
                val adapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item)
                //選択結果に応じてSpinnerの選択肢を変える
                when (which) {
                //アルファベット
                    0 -> {
                        val ALPHABET_SIZE = 'Z' - 'A'
                        var alphabet = 'A'
                        for (i in 0..ALPHABET_SIZE) {
                            adapter.add(alphabet++.toString())
                        }
                    }
                //数字
                    1 -> (0..9).forEach { i -> adapter.addAll(i.toString()) }
                //制御キー
                    2 -> adapter.addAll("Backspace", "Enter", "Shift", "Ctrl", "Alt", "Pause", "Space", "PageUp", "PageDown", "End", "Home", "←", "↑", "→", "↓", "PrintScreen", "Insert", "Delete", "Win", "NumLock", "ScrollLock", "Esc", "Tab")
                //ファンクションキー
                    3 -> (1..12).forEach { i -> adapter.add("F" + i.toString()) }
                    4 -> adapter.addAll("VolUp", "VolDown", "VolMute")
                //記号
                //4 -> adapter.addAll(":", ";", "+", ",", "-", "=", ".", "/", "@", "[", "\\", "]", "^", "_")
                }
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                spinner.adapter = adapter
                spinner.setSelection(0)

                val removeButton = addKeyCodeView.findViewById(R.id.removeButton) as Button
                removeButton.setOnClickListener { view ->
                    val parent = view.parent as ViewGroup
                    keyCodeViewList.remove(parent)
                    parent.removeAllViews()
                }
                keyCodesLayout.addView(addKeyCodeView)
                keyCodeViewList.add(addKeyCodeView)
            }
            builder.show()
        }
    }

    companion object {
        fun newInstance(emptySpace: Int): AddKeyDialogFragment {
            val fragment = AddKeyDialogFragment()
            val args = Bundle()
            args.putInt("emptySpace", emptySpace)
            fragment.arguments = args
            return fragment
        }
    }
}
